Hex floats

LeventErkok opened the pull request:
This is a proposal to add support for hexadecimal floating-point literals.

[Rendered](http://github.com/LeventErkok/ghc-proposals/blob/hexFloats/proposals/0000-hexFloats.rst)

bitemyapp commented:
Apologies if this is a dumb question, but I couldn't find a certain answer in the proposal.

Under, "Effect and Interactions", you say:

>None. The addition is orthogonal, and the changes to the grammar is unambiguous by design. No significant complexity to any part of the compiler anticipated.

Have you tested the additions to the grammar in GHC yet? I think it should work fine as you describe, but I was wondering. Seems like it would be isolated to the parsing of numerical literals.

LeventErkok commented:
@bitemyapp No I have not tested in the GHC grammar. But as you said it should be isolated to parsing of the floating-point numeric literals, and the maximal-munch rule should guarantee no conflicts. (The starting "0x" and the "p" are required, and the syntax is designed this way to ensure there's no ambiguity.)

To wit, the parser I wrote for it in the HexFloats package (https://github.com/LeventErkok/FloatingHex/blob/master/Data/Numbers/FloatingHex.hs#L63) is really a straightforward scan of the string. Should be easier with the yacc based parser in GHC.

bitemyapp commented:
Thank you for elaborating!

cartazio commented:
hex format floats are the only printed format specified by the ieee
floating standard. They are also the only format that round trips according
to the ieee spec

I'll take a look at this proposal myself and make sure the two reconcile :)


And perhaps see about sharing the relevant language.
On Tue, Jan 17, 2017 at 6:37 PM Chris Allen <notifications@github.com>
wrote:

> Thank you for elaborating!
>
>
>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ghc-proposals/ghc-proposals/pull/37#issuecomment-273336163>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQwiNLPals7UniDc_oR4B8-0OGr4AWks5rTVCbgaJpZM4LmSpU>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>

LeventErkok commented:
The pretty printer I wrote for for this, called `showHFloat` (https://github.com/LeventErkok/FloatingHex/blob/master/Data/Numbers/FloatingHex.hs#L68-L78) indeed should guarantee round-tripping; i.e., the following should hold:

         for any non-negative float d except NaN and infinity:
               parseHFloat (showHFloat d) = d

Note that there's no way to write `NaN` or `Infinity` in this notation. Also, the parser only has to deal with positive quantities, so, the above also requires `d >= 0`, and negative zero is not-allowed either by the same token. (Negative quantities are not an issue, as they are simply multiplication by `(-1)` and should not be parsed as part of the literal.)

Round-tripping the other way isn't possible, since the same float can be written in multiple equivalent ways using the hexadecimal notation. The canonical example is `0`: All of the following: `0x0pN`, for any `N`, is equivalent to  `0`.

Having said all this: It would indeed be great to get a second pair of eyes to make sure what I claimed above is actually true of the code I wrote! Floating point is tricky, and I might've introduced a bug!

nomeata commented:
Should this also extend the `Read` instance for floating point numbers? Maybe yes, as `Read` is supposed to read valid source code. But then the implications might be greater.

LeventErkok commented:
@nomeata Great point! Yes, I think "read" instance should be extended as well. I've updated the proposal to make a point of that.

I don't see why that would increase complexity, however. It's rather a straightforward parse: https://github.com/LeventErkok/FloatingHex/blob/master/Data/Numbers/FloatingHex.hs#L68-L78 can be a starting point. That code would need to be merged with the existing read instance, but since we're starting with `0x` there should be no ambiguity.

nomeata commented:
Well, it is a change that cannot be guarded by a language pragma. And we would have to worry about breaking code that for some reason expects `read` to partially parse the `0` and then does something else with the `x`. (Maybe unlikely, but worth debating.)

There was a similar discussion recently about changes in behavior of `read`, maybe @treeowl can chime in here with general advicse?

LeventErkok commented:
@nomeata Agreed. It's a shame that it cannot be guarded by a language pragma.

I agree, this would break backwards compatibility in those rare cases. I wonder how often that sort of parsing is done? I personally never recall writing such code, but my experience is certainly not the deciding factor here.

rwbarton commented:
I'm okay with this as long as it is restricted to ordinary finite floating-point values, which it sounds like it is.

> decimal notation is not sufficient for expressing floats precisely

This is not really true: any ordinary finite floating-point value has a value that can be exactly expressed as a terminating decimal expansion. (Granted, that decimal expansion might be ~1000 digits long.)

LeventErkok commented:
@rwbarton  Regarding the precision: You are correct, I've updated the text of the proposal to reflect the fact that writing floats precisely in decimal might require lots of digits.

One aside: The format actually implicitly allows for infinity, if the specified number is beyond what the underlying representation can support. (Consider `0x1p5000`).

I think the right thing to do when the literal is too large is to print a warning, similar to this:

```
Prelude Data.Word> 200000::Word16

<interactive>:3:1: warning: [-Woverflowed-literals]
    Literal 200000 is out of the Word16 range 0..65535
3392
```

But still convert it to `Infinity` as its value. I've added this to the proposal as well.

rwbarton commented:
I don't think this last part about `Infinity` is exactly right, but first a question: What exactly do any of your examples desugar to?

LeventErkok commented:
@rwbarton I'm not sure what you mean by "desugar?" They all become Floats or Double's, just like what you would get if you would write `2.5`. Maybe I'm misunderstanding your question.

rwbarton commented:
Floating-point literals are overloaded, so `2.5` means `fromRational (25 % 10)`. It could have any type that is an instance of `Fractional`, which is more than just `Float` and `Double`, for example `Rational` itself. I'm assuming that this new syntax is supposed to desugar to things like `fromRational (173 % 256)`.

There is no concept of "infinity" in `Fractional`, so you cannot write infinity as a floating-point literal. You can of course write the IEEE double value `Infinity` if you happen to know that `Double` is represented by IEEE doubles on your system and you choose a large enough number and the type of your literal is actually `Double`; this is true already today.

LeventErkok commented:
@rwbarton I see. I would suggest that such literals are not overloaded; and simply have the type `RealFloat a => a` instead. This is really about machine floats we are talking about. Updated the proposal accordingly.

yav commented:
Is there really a problem with infinity?   I think that we could simply translate the notation via Rational as usual---if a number is too big, then it would be up to the instance to choose how to represent it.  The current instances already appear to be translating numbers that are too big to infinity.  For example:

    > fromRational (10^64 % 1) :: Float
    Infinity

LeventErkok commented:
@yav You are correct; there isn't really an issue with `Infinity`. But the text was confusing, and so was I!

The only issue with "overflow" is weather GHC should print a warning about it, and indeed the referenced standard recommends that.

I updated the proposal to clarify the point about `Infinity` and printing warnings about overflow with a proper citation. I hope it reads better now.

rwbarton commented:
I think it's important to be simple and consistent and desugar hexadecimal floating literals to the same `Fractional a => a` type as ordinary floating literals. After all we let the user write large `Integer` literals in hexadecimal even though hexadecimal integer literals are mostly used for machine-word-sized or smaller things.

I also think we should treat an out-of-`Double`-range literal `0x1p5000 :: Double` the same way as we treat the existing literal `1e5000 :: Double`. If we should give a warning for the former then we should probably also give a warning for the latter; it's an orthogonal issue to this proposal.

gracjan commented:
As with any syntax change I would appreciate inclusion of people that write indenpendant haskell parsers, for example syntax coloring for emacs/vim/atom/vscode/whatever. Some syntaxes produce more issues than necessary, as example I can point to https://github.com/haskell/haskell-mode/issues/1185.

Having said that I do not see any problems with above spec, I would appreciate confirmation from @cartazio, that we are copying bit-for-bit syntax from IEEE spec.

LeventErkok commented:
@rwbarton Agreed. Cleaned-up the proposal slightly to keep the desugaring as is, and emphasized the warning for underflow/overflow cases along with the usual decimal-floats case; which is indeed orthogonal to this proposal.

yav commented:
I would like to propose that we deviate from the standard in the following two ways:
  * Skip the optional suffix (`f` or `l`).  This is used to specify the size of the literals, but in Haskell we resolve this in using the class system, so there's no need for it.
  * Allow the exponent (the `p` part) to be optional.

These same changes are made in other language (e.g. Lua).  In the standard and C, the exponent is not optional to avoid ambiguity with the `F` suffix (i.e., is this a hex digit, or an actual suffix).  Since for us the suffix doesn't really make sense, I think it is reasonable to make the exponent optional, like in the decimal notation.

LeventErkok commented:
@yav Regarding deviations:

  * Dropping the suffix is indeed the right thing to do for Haskell. (My library implementation http://hackage.haskell.org/package/FloatingHex indeed drops the suffix.) I added that to the proposal.

   * I'm not sure if making the exponent optional is a good idea. Note that the decimal point itself is optional, you can write `0x10p0`, for instance, to mean `16`. If you drop the exponent, you'd end up with `0x10` which looks a lot like an integral value! Of course, it still means `16`; but perhaps keeping the exponent does serve a purpose in that it visually indicates a float, the whole point of the notation in the first place.

Dropping the exponent might still be OK though, I'm not too attached to it.

rwbarton commented:
Making the exponent optional would be consistent with standard Haskell floating literals.
```
Prelude> :t 16
16 :: Num a => a
Prelude> :t 16.0
16.0 :: Fractional a => a
Prelude> :t 16e0
16e0 :: Fractional a => a
Prelude> :t 16.

<interactive>:1:4:
    parse error (possibly incorrect indentation or mismatched brackets)
```

LeventErkok commented:
@rwbarton @yav 

Sounds good. I added clarifying text to mention `p` is optional. Also added text on desugaring; i.e., if we see a *dot* or *p* then we desugar through `fromRational` otherwise we desugar through `fromInteger`, following the usual rules.

yav commented:
Yesterday I had a go at actually implementing this in GHC.  I haven't finished it yet, but it looks like it should be fairly straight-forward.  In the process, another small thing came up:

According to the standard (and Lua), literals like this are allowed:  `0xFF.`  and `0x.FF`.  Haskell does not allow such literals in the decimal floating point location (e.g., no `5.` or `.5`), so I'd vote that we don't do them in the hex notation either.  So another deviation from the standard would be that:

   * if a literal contains a dot, then it must have some hex digits before and after it

Thoughts?

LeventErkok commented:
@yav Good point. The deviation with `.5` and `5.` is already there, so it's probably best to keep consistency here.

I've updated the text to note this exception.

cartazio commented:
it seems like the discussion so far is consistent with IEEE floats 2008 standard

it looks like the referenced C standard is merely an adaptation of the IEEE language?
i've also attached the ieee pdf below along with relevant excerpted text 
[IEEE 754 2008 floating point standard fpstandard-2008.pdf](https://github.com/ghc-proposals/ghc-proposals/files/739880/IEEE.754.2008.floating.point.standard.fpstandard-2008.pdf)


```
5.12.3 External hexadecimal-significand character sequences representing finite numbers 5.12.3.0
Language standards should provide conversions between all supported binary formats and external hexadecimal-significand character sequences. External hexadecimal-significand character sequences for finite numbers shall be described by the following grammar, which defines a hexSequence:
 sign [+ −]
digit [0123456789]
hexDigit [0123456789abcdefABCDEF]
hexExpIndicator [Pp]
hexIndicator "0" [Xx]
hexSignificand ( {hexDigit} * "." {hexDigit}+ | {hexDigit}+ "." | {hexDigit}+ )
decExponent   {hexExpIndicator} {sign}? {digit}+
hexSequence  hexSequence

where each line is a name followed by a rule in which ‘[...]’ selects one of the terminal characters listed between the brackets, ‘{...}’ refers to an earlier named rule, ‘(... | ... | ...)’ indicates a choice of one of three alternatives, straight double quotes enclose a terminal character, ‘?’ indicates that there shall be either no instance or one instance of the preceding item, ‘*’ indicates that there shall be zero or more instances of the preceding item, and ‘+’ indicates that there shall be one or more instances of the preceding item.
The hexSignificand is interpreted as a hexadecimal constant in which each hexDigit represents a value in the range 0 through 15 with the letters ‘a’ through ‘f’ representing 10 through 15, regardless of case. Within the hexSignificand, the first (leftmost) character is the most significant. If present, the period defines the start of a hexadecimal fractional part; if the period is to the right of all hexadecimal digits the hexSignificand is an integer. The decExponent is interpreted as an optionally-signed integer expressed in decimal following the hexExpIndicator, again with the most significant digit first.
The value of a hexSequence is the value of the hexSignificand multiplied by two raised to the power of the value of the decExponent, negated if there is a leading ‘−’ sign. The hexIndicator and the hexExpIndicator have no effect on the value.
When converting to hexadecimal-significand character sequences in the absence of an explicit precision specification, enough hexadecimal characters shall be used to represent the binary floating-point number exactly. Conversions to hexadecimal-significand character sequences with an explicit precision specification, and conversions from hexadecimal-significand character sequences to supported binary formats, are correctly rounded according to the applicable binary rounding-direction attribute, and signal all exceptions appropriately.

```

LeventErkok commented:
@cartazio Yes, the text seems to be the same. @yav suggested we make the `p` part optional, and we do require a digit both before and after the dot if the dot is indeed present; following other Haskell floating-point literals; see earlier comments.

These deviations are documented in the proposal.

cartazio commented:
its worth noting that the IEEE document doesn't mention type suffixes on the literals, thats strictly a C-ism

cartazio commented:
So what desugaring would it be in the no `p` or `P` case? 

`fromInteger someLiteral` ?

cartazio commented:
so would enabling hexfloating points imply enabling hex integers?

cartazio commented:
also I would like to note / emphasise that the proposal would have fewer exceptions if we talk about the delta from the IEEE spec as the base truth, rather than the C standard

yav commented:
@cartazio I am not sure what you mean?   Hex integers are just part of Haskell.

The idea here is that if a literal has `p` or `P` in it, then it will be de-sugared using `fromRational`, even if it has no dots in it.   This is basically the same as the decimal notation, where literals contains `e` or `E` are similarly translated using `fromRational`.  For example:

    > :t 1e10
    1e10 :: Fractional t => t

rwbarton commented:
The proposal can also be understood as adding the bold entries to the table below, in which the non-bold entries are the existing ways to write decimal and hexadecimal numeric literals in Haskell.

decimal | hexadecimal | desugaring
------- | ----------- | ----------
`10` | `0xa` | fromInteger 10
`10.0` | **`0xa.0`** | fromRational (10 % 1)
`1e1` | **`0x5p1`** | fromRational (10 % 1)
`1.0e1` | **`0x5.0p1`** | fromRational (10 % 1)

Relative to the 2010 Report (section [Numeric Literals](https://www.haskell.org/onlinereport/haskell2010/haskellch2.html#x7-190002.5)), we add the following productions:

pexponent 	→ 	(p | P) [+ | -] decimal
float 	→ 	0x hexadecimal . hexadecimal [pexponent] | 0X hexadecimal . hexadecimal [pexponent]
.......... | 	0x hexadecimal pexponent | 0X hexadecimal pexponent

GHC implements two related extensions, but the interactions are obvious.

* [`NegativeLiterals`](http://ghc.readthedocs.io/en/latest/glasgow_exts.html#negative-literals): Add `-LIT` as a numeric literal, where `LIT` is any of the above existing numeric literals.

* [`NumDecimals`](http://ghc.readthedocs.io/en/latest/glasgow_exts.html#fractional-looking-integer-literals): Whether to desugar a literal in terms of `fromInteger` or `fromRational` is determined by whether its mathematical value is an integer or not, rather than whether it was syntactically written with a `.`, `e` or `p`.

cartazio commented:
ok, this all make sense, i'm in agreement / support then :)

On Mon, Jan 30, 2017 at 7:50 PM, Reid Barton <notifications@github.com>
wrote:

> The proposal can also be understood as adding the bold entries to the
> table below, in which the non-bold entries are the existing ways to write
> decimal and hexadecimal numeric literals in Haskell.
> decimal hexadecimal desugaring
> 10 0xa fromInteger 10
> 10.0 *0xa.0* fromRational (10 % 1)
> 1e1 *0x5p1* fromRational (10 % 1)
> 1.0e1 *0x5.0p1* fromRational (10 % 1)
>
> Relative to the 2010 Report (section Numeric Literals
> <https://www.haskell.org/onlinereport/haskell2010/haskellch2.html#x7-190002.5>),
> we add the following productions:
>
> pexponent → (p | P) [+ | -] decimal
> float → 0x hexadecimal . hexadecimal [pexponent] | 0X hexadecimal .
> hexadecimal [pexponent]
> .......... | 0x hexadecimal pexponent | 0X hexadecimal pexponent
>
> GHC implements two related extensions, but the interactions are obvious.
>
>    -
>
>    NegativeLiterals
>    <http://ghc.readthedocs.io/en/latest/glasgow_exts.html#negative-literals>:
>    Add -LIT as a numeric literal, where LIT is any of the above existing
>    numeric literals.
>    -
>
>    NumDecimals
>    <http://ghc.readthedocs.io/en/latest/glasgow_exts.html#fractional-looking-integer-literals>:
>    Whether to desugar a literal in terms of fromInteger or fromRational
>    is determined by whether its mathematical value is an integer or not,
>    rather than whether it was syntactically written with a ., e or p.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ghc-proposals/ghc-proposals/pull/37#issuecomment-276240177>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQwj2pd4vcLKSakqfUoU-pv5uPnRuUks5rXoVLgaJpZM4LmSpU>
> .
>

yav commented:
I made a patch implementing this.  If you are interested, have a look here:

https://phabricator.haskell.org/D3066

LeventErkok commented:
@yav Thanks Iavor.. Two questions:

   

-  `readHexRational` seems to implement reading negative numbers. Is that really necessary? The problem here is that it doesn't correctly work for "-0x0p1": This is the '-0' literal (which is different than +0), but so far as I can see it'll be lost in translation. I was hoping that this would never be an issue, you'd only "read" positive literals, and `-` would be multiplication by `-1`, which will do the correct thing. 

-    We also need a printer: `showHFloat`, which should go into `module Numeric`. I have an implementation of that here: http://hackage.haskell.org/package/FloatingHex-0.4/docs/src/Data-Numbers-FloatingHex.html#showHFloat which I think you can use verbatim. (Though please go over it in case I screwed up something!)

yav commented:
The parsing of the negative literals is only for interacting with the `NegativeLietarals` extension.  It doesn't seem that the decimal float negative literal notation knows about `-0` either, so I think decimal and hex should work the same way. 

Ah, yes, I forgot about `showHFloat`, I'll add this.

I haven't looked into adding the warning---I think maybe we should just make a normal GHC ticket for that, and add it to all floating point literals.

rwbarton commented:
On the subject of `-0` under `NegativeLiterals`, see https://ghc.haskell.org/trac/ghc/ticket/13211.

LeventErkok commented:
@yav Agreed; overflow/undeflow deserves its own ticket. Let me know if you want me to file that.

LeventErkok commented:
Overflow/undeflow ticket filed for GHC: http://ghc.haskell.org/trac/ghc/ticket/13232#ticket

yav commented:
I think we've worked out the details of this proposal.  @LeventErkok do you think it is ready to be marked as "For committee consideration"?  If so, could you please re-label it.  I don't mean to be pushy, it is just that I just learned about this next step in the process, and I wasn't sure if you were aware of it.

On the implementation front, I've been busy, but basically I just need to do pretty printing function (for Numeric) and add the parsing functions to the Read instance.   The pretty printing is a bit more complex than I thought, because we have to implement all the flavors the `F`, `E`, and `G` versions.  It is not too bad, just many small  cases to consider, and I haven't had the time to do it properly.

LeventErkok commented:
@yav Agreed! I think it's ready for committee consideration. I added a link to your patch from the proposal text itself.

For the label: I do not know how to add/change labels. (I suppose I don't have the right privileges to do so?) Last time @nomeata added the "Under discussion" label, perhaps he can do the change?

Regarding the pretty printer: I don't understand your comment about `F`, `E`, `G` versions. I suppose you mean the `showFFloat`, `showEFloat` and `showGFloat` functions that already exist. Why do they need a change? My expectation would be that they'd remain unchanged, and we'd simply add a `showHFloat` function that I already have the code for here: http://hackage.haskell.org/package/FloatingHex-0.4/docs/src/Data-Numbers-FloatingHex.html#showHFloat

Why do we need to change the others?

yav commented:
I just meant that we probably want to add the parallel `showHexFFloat`, `showHexEFloat`, and `showHexGFloat`.   The one in your package is `showHexEFloat Nothing`, I think.  In GHC all of the above are implemented in terms of a single function with many parameters, and it already implements a bunch of stuff that we can reuse (e.g., rounding) so it shouldn't be too hard.

The main weirdness with hex floats is that the base of the digits and the base of the exponent differ (16 and 2 respectively), but fortunately a lot of the existing code is already parametrized to work for any base, so it can be reused.

yav commented:
In my view, there is a `Labels` menu in the top-right part of the page---perhaps it is a permission issue.  I just relabeled it for review by the committee.

LeventErkok commented:
I don't see why we need `showHexFFloat`.. etc. Iavor. The whole point of the format is that you do *not* need rounding; it's a precise "textual" representation of the value. So the width field is irrelevant. And the `E`, `F`, `G` is precisely about how the string looks like; which is again dictated by the hex-floats themselves. It'll always be `0xAB.CDEpNNN`.

Another way to think of this is that we're really adding a new `H` class to the existing `F`, `G`, and `E` formats for pretty printing purposes. The other ones remain exactly as they are.

yav commented:
I was just thinking that sometimes you may want to see the number printed using the fully expanded hex digits  without any exponents (I guess that would be the `F` behavior) and the "rounding" could be useful if you wanted numbers padded with extra 0s for showing them in nice tables.

However, I am starting to think that perhaps this kind of functionality belongs more in a pretty printing library rather than `Numeric`, so I'd be fine with just adding `showHFloat`---it is much simpler anyway :)

LeventErkok commented:
Padding might be useful, but I do agree that it is probably best to handle that in a library.

nomeata commented:
This was actually merged in a9f92c43ab7beea76ef6f3e3f0f4c1ee9a7546d6 (but GitHub did not notice, it seems).
